src/textinput/TextInput.ts

```diff
@@ -0,0 +1,162 @@
+
+@theme(css)
+export default class TextInput extends TextInputBase<TextInputProperties> {
+	onBlur (event: FocusEvent) { this.properties.onBlur && this.properties.onBlur(event); }
+	onChange (event: Event) { this.properties.onChange && this.properties.onChange(event); }
+	onClick (event: MouseEvent) { this.properties.onClick && this.properties.onClick(event); }
+	onFocus (event: FocusEvent) { this.properties.onFocus && this.properties.onFocus(event); }
+	onInput (event: Event) { this.properties.onInput && this.properties.onInput(event); }
+	onKeyDown (event: KeyboardEvent) { this.properties.onKeyDown && this.properties.onKeyDown(event); }
+	onKeyPress (event: KeyboardEvent) { this.properties.onKeyPress && this.properties.onKeyPress(event); }
+	onKeyUp (event: KeyboardEvent) { this.properties.onKeyUp && this.properties.onKeyUp(event); }
+	onMouseDown (event: MouseEvent) { this.properties.onMouseDown && this.properties.onMouseDown(event); }
+	onMouseUp (event: MouseEvent) { this.properties.onMouseUp && this.properties.onMouseUp(event); }
+	onTouchStart (event: TouchEvent) { this.properties.onTouchStart && this.properties.onTouchStart(event); }
+	onTouchEnd (event: TouchEvent) { this.properties.onTouchEnd && this.properties.onTouchEnd(event); }
+	onTouchCancel (event: TouchEvent) { this.properties.onTouchCancel && this.properties.onTouchCancel(event); }
+
+	render() {
+		const {
+			describedBy,
+			disabled,
+			formId,
+			invalid,
+			label,
+			maxLength,
+			minLength,
+			name,
+			placeholder,
+			readOnly,
+			required,
+			type = 'text',
+			value
+		} = this.properties;
+
+		const stateClasses = [
+			disabled ? css.disabled : null,
+			invalid ? css.invalid : null,
+			invalid === false ? css.valid : null,
+			readOnly ? css.readonly : null,
+			required ? css.required : null
+		];
+
+		const textinput = v('div', { classes: this.classes(css.inputWrapper) }, [
+			v('input', {
+				bind: this,
+				classes: this.classes(css.root),
+				'aria-describedby': describedBy,
+				disabled,
+				'aria-invalid': invalid + '',
+				maxlength: maxLength ? maxLength + '' : null,
+				minlength: minLength ? minLength + '' : null,
+				name,
+				placeholder,
+				readOnly,
+				'aria-readonly': readOnly ? 'true' : null,
+				required,
+				type,
+				value,
+				onblur: this.onBlur,
+				onchange: this.onChange,
+				onclick: this.onClick,
+				onfocus: this.onFocus,
+				oninput: this.onInput,
+				onkeydown: this.onKeyDown,
+				onkeypress: this.onKeyPress,
+				onkeyup: this.onKeyUp,
+				onmousedown: this.onMouseDown,
+				onmouseup: this.onMouseUp,
+				ontouchstart: this.onTouchStart,
+				ontouchend: this.onTouchEnd,
+				ontouchcancel: this.onTouchCancel
+			})
+		]);
+
+		let textinputWidget;
+
+		if (label) {
+			textinputWidget = w(Label, {
+				classes: this.classes(css.label, ...stateClasses),
+				formId,
+				label
+			}, [ textinput ]);
+		}
+		else {
+			textinputWidget = v('div', {
```

Comment:
As raised by @mwistrand on another pr, do you think this should be a `span` rather than a `div` to reflect the inline nature of the equivalent dom element?